controllers/SiteController.php

```diff
@@ -167,4 +172,132 @@ private function checkIfUserMatchToToken($token)
+
+        if($form->load($_POST) && $form->validate()) {
+            $this->forgetPassword($form);
+        }
+
+        return $this->render('forgotten-password', [
+            'forgottenPasswordModel' => $form,
+        ]);
+    }
+
+    /**
+     * @param $form
+     * @throws \Exception
+     */
+    private function forgetPassword($form)
+    {
+        $user = $this->checkIfUserMatchToEmail($form->email);
+
+        $tokenTry = 0;
+        do {
+            if($tokenTry === 10) {
+                throw new CannotCreateTokenException();
+            }
+            $token = $this->createUserToken();
+            $tokenTry++;
+        } while (!is_null(User::findOne(['token' => $token])));
+
+        $user->token = $token;
+
+        if (!$user->save()) {
+            throw new CannotSaveException($user);
+        }
+        $this->mailToUserPasswordForgotten($user);
+
+        return $this->redirect("/");
+    }
+
+    /**
+     * @param null $token
+     * @return string
+     * @throws \yii\base\Exception
+     * @throws CannotSaveException
+     */
+    public function actionChangePassword($token = null)
+    {
+        $user = $this->checkIfUserMatchToToken($token);
+
+        $form = new ChangePasswordForm();
+
+        if($form->load($_POST) && $form->validate()) {
+            $this->changePassword($form, $user);
+        }
+
+        return $this->render('change-password', [
+            'model' => $form,
+        ]);
+    }
+
+    /**
+     * @param $form
+     * @param $user
+     * @throws CannotSaveException
+     * @throws \yii\base\Exception
+     */
+    private function changePassword($form, $user)
+    {
+        $user->password_hash = Yii::$app->security->generatePasswordHash($form->password);
+        $user->token = null;
+
+        if (!$user->save()) {
+            throw new CannotSaveException($user);
+        }
+
+        Yii::$app->user->login($user);
+        $_SESSION["userPasswordHash"] = $user->password_hash;
+        return $this->redirect("/");
+    }
+
+    /**
+     * @param $email
+     * @return User|null
+     * @throws \Exception
+     */
+    private function checkIfUserMatchToEmail($email)
+    {
+        $notFoundException = NotFoundHttpException::class;
+        if (!is_null($user = User::findOne(['email' => $email]))) {
+            return $user;
+        }
+
+        throw new $notFoundException();
+    }
+
+    /**
+     * @param User $user
+     * @return bool
+     * @throws \PHPMailer\PHPMailer\Exception
+     * @throws CannotSendMailException
+     */
+    private function mailToUserPasswordForgotten(User $user)
+    {
+        $mail = Util::getConfiguredMailerForMailhog();
+        $mail->addAddress($user->email);
+        $mail->isHTML(false);
+        $mail->CharSet = 'UTF-8';
+
+        $mail->Subject = "Forgotten your Collab'media account password";
+        $mail->Body = 'Click on the following link to reset your password : http://127.0.0.1:8000/site/change-password/' . $user->token ;
```

Comment:
What will happen when we will run the app on our final domain ? What will happen if I have Docker Toolbox and I the IP address of my project is not localhost ?

Please create a public const 'APP_BASE_URL' in Util component, and use this constant everywhere you need an absolute URL.

Reply:
I fully agree